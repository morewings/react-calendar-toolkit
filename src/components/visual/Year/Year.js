import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import classes from './Year.module.css';

const Year = ({
  isHighlighted,
  onDateSet,
  date,
  isSelected,
  isCurrent,
  name,
}) => {
  const handleClick = () => {
    onDateSet(date);
  };

  return (
    <div
      tabIndex="0"
      role="button"
      onClick={handleClick}
      onKeyPress={handleClick}
      className={classNames({
        [classes.wrapper]: true,
        /** Conditional class to display, if year belongs to same year as today */
        [classes.isCurrent]: isCurrent,
        /** Conditional class to display, if year is selected */
        [classes.isSelected]: isSelected,
        /** Conditional class to display, if year is highlighted */
        [classes.isHighlighted]: isHighlighted,
      })}
      aria-label={name.numeric}>
      {name.numeric}
    </div>
  );
};

Year.propTypes = {
  /** Flag, showing if this year entry is __selected__. */
  isSelected: PropTypes.bool.isRequired,
  /** Flag, showing if this year entry is __highlighted__. */
  isHighlighted: PropTypes.bool.isRequired,
  /** Function to set date for Calendar, e. g. `onDateSet(date)`. */
  onDateSet: PropTypes.func.isRequired,
  /** Date for this year entry */
  date: PropTypes.instanceOf(Date).isRequired,
  /** Flag, showing if this month entry and current date __belong to the same year__. */
  isCurrent: PropTypes.bool.isRequired,
  /** Object, which has `numeric` property with current year number. */
  name: PropTypes.shape({
    numeric: PropTypes.number.isRequired,
  }).isRequired,
};

export default Year;
