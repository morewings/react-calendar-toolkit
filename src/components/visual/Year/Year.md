`Year` component displays specific date with __year precision__. `onDateSet` prop allows to set this date as selected. Here is how `Datepicker` with custom `Year` component looks like:

```js
import DatePicker from 'react-calendar-toolkit';

const style = {
  boxSizing: 'border-box',
  flex: '1 1 calc(100% / 5 - 4px)',
  color: 'black',
  textAlign: 'center',
  padding: '0',
  border: '2px dashed pink',
  margin: '6px 2px',
  height: '36px',
  lineHeight: '36px'
};

const CustomComponent = ({
  isHighlighted,
  onDateSet,
  date,
  isSelected,
  isSameYear,
  name,
}) => {
  return (
    <div style={style}>
      {name.numeric}
    </div>
  )
};

<DatePicker
  renderYearAs={CustomComponent}
  minPrecision="year"
  startDate={new Date(1999, 0, 1)}
  endDate={new Date(2020, 1, 25)}
  title="Demo datepicker" />
```

### Year Component example
```js { "file": "../Year.js" }
```

### Wrap Year Calendar with custom component
You can provide wrapper component to Year Calendar, which can be React Component or DOM node.

```js
import DatePicker from 'react-calendar-toolkit';

const style = {
  display: 'flex',
  width: '276px',
  flexDirection: 'row',
  flexWrap: 'wrap',
  padding: '0 8px 8px',
  margin: '0 8px 8px',
  height: '260px',
  border: '2px dashed pink',
  overflow: 'auto',
};

const CustomComponent = ({children}) => (
  <div style={style}>{children}</div>
);

<DatePicker
  minPrecision="year"
  wrapYearWith={CustomComponent}
  startDate={new Date(1999, 0, 1)}
  endDate={new Date(2020, 1, 25)}
  title="Demo datepicker" />
```


