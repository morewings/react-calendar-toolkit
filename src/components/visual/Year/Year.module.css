.scrollContainer{
  height: var(--gridHeight);
  overflow: auto;
  padding: 0 0 var(--innerPadding);
}

.grid {
  display: flex;
  width: var(--innerWidth);
  flex-direction: row;
  flex-wrap: wrap;
  padding: 0 var(--innerPadding) var(--innerPadding);
}

.wrapper {
  text-align: center;
  box-sizing: border-box;
  flex: 1 1 calc(100% / 5 - 2px);
  height: var(--dayHeight);
  line-height: var(--dayHeight);
  border: 1px solid transparent;
  cursor: pointer;
  border-radius: var(--borderRadius);
  margin: var(--innerPadding) 1px 0;
  color: var(--baseColor);
  user-select: none;
}

.wrapper:active,
.wrapper:focus {
  outline: none;
}

.wrapper:focus:not(.isSelected),
.wrapper:hover{
  background-color: var(--lightGray);
  color: var(--baseColor);
}

.isSelected {
  color: white;
  background-color: var(--active);
}

.isCurrent {
  border-color: var(--gray);
}