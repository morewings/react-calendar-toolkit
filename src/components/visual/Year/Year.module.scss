@import "./../../../variables";

.scrollContainer{
  height: $gridHeight;
  overflow: auto;
}

.grid {
  display: flex;
  width: $innerWidth;
  flex-direction: row;
  flex-wrap: wrap;
  padding: 0 $innerPadding $innerPadding;
}

.wrapper {
  text-align: center;
  box-sizing: border-box;
  flex: $yearCellFlex;
  height: $dayHeight;
  line-height: $dayHeight;
  border: 1px solid transparent;
  cursor: pointer;
  border-radius: $borderRadius;
  margin-top: $innerPadding;
  &:active{
    outline: none;
  }
  &:hover{
    background-color: $lightGray;
  }
  &.isSelected {
    color: white;
    background-color: $active;
  }
  &.isSameYear {
    border-color: $gray;
  }
}