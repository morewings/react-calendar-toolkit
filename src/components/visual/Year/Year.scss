@import "./../../../variables";

.year-scroll-container{
  height: $gridHeight;
  overflow: auto;
}

.year-grid-wrapper {
  display: flex;
  width: $innerWidth;
  flex-direction: row;
  flex-wrap: wrap;
  padding: 0 $innerPadding $innerPadding;
}

.year-wrapper {
  text-align: center;
  box-sizing: border-box;
  flex: $yearCellFlex;
  height: $dayHeight;
  line-height: $dayHeight;
  border: 1px solid transparent;
  cursor: pointer;
  border-radius: $borderRadius;
  margin-top: $innerPadding;
  &:active{
    outline: none;
  }
  &:hover{
    background-color: $lightGray;
  }
  &.disabled {
    color: $disabled;
  }
  &.isSelected {
    color: white;
    background-color: $active;
  }
  &.isSameYear {
    border-color: $gray;
  }
}