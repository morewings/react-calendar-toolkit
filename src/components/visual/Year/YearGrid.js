import React from 'react';
import PropTypes from 'prop-types';
import classes from './Year.module.css';

const YearGrid = props => (
  <div className={classes.scrollContainer}>
    <div className={props.className}>{props.children}</div>
  </div>
);

YearGrid.propTypes = {
  className: PropTypes.string,
  children: PropTypes.node.isRequired,
};

YearGrid.defaultProps = {
  className: classes.wrapper,
};

export default YearGrid;
