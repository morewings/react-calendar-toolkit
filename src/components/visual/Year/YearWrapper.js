import React from 'react';
import PropTypes from 'prop-types';

const YearWrapper = props => (
  <div className="year-scroll-container">
    <div className={props.className}>{props.children}</div>
  </div>
);

YearWrapper.propTypes = {
  className: PropTypes.string.isRequired,
  children: PropTypes.node.isRequired,
};

export default YearWrapper;
