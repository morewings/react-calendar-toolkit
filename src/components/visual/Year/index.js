export {default} from './Year';
export {default as YearWrapper} from './YearWrapper';
