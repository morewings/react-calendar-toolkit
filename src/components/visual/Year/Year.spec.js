import React from 'react';
import {DatepickerMockProvider as Provider} from 'utils/testProvider';
import {render, fireEvent} from '@testing-library/react';
import Year from './Year';

const defaultProps = {
  onDateSet: jest.fn(),
  date: new Date(2020, 0, 12),
  name: {
    numeric: 1,
  },
};

describe('Year', () => {
  beforeEach(() => {
    defaultProps.onDateSet.mockClear();
  });

  describe.each`
    belongsToSameMonth | isCurrent | isSelected | isHighlighted | isWeekend
    ${true}            | ${true}   | ${true}    | ${true}       | ${true}
    ${false}           | ${true}   | ${true}    | ${true}       | ${true}
    ${false}           | ${false}  | ${false}   | ${false}      | ${false}
  `(
    'display props',
    ({
      belongsToSameMonth,
      isCurrent,
      isSelected,
      isDisabled,
      isHighlighted,
      isWeekend,
    }) => {
      it(`it renders`, () => {
        const props = {
          ...defaultProps,
          belongsToSameMonth,
          isCurrent,
          isSelected,
          isDisabled,
          isHighlighted,
          isWeekend,
        };
        const {asFragment} = render(<Year {...props} />, {
          wrapper: ({children}) => <Provider>{children}</Provider>,
        });
        expect(asFragment()).toMatchSnapshot();
      });
    }
  );

  it('handles click when belongsToSameMonth === true', () => {
    const props = {
      ...defaultProps,
      belongsToSameMonth: true,
      isCurrent: false,
      isSelected: false,
      isDisabled: false,
      isHighlighted: false,
      isWeekend: false,
    };
    const {getByRole} = render(<Year {...props} />, {
      wrapper: ({children}) => <Provider>{children}</Provider>,
    });
    fireEvent.click(getByRole('button'));
    expect(defaultProps.onDateSet).toHaveBeenCalledTimes(1);
    expect(defaultProps.onDateSet).toHaveBeenCalledWith(defaultProps.date);
  });
});
