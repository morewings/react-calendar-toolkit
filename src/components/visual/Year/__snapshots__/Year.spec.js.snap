// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Year display props it renders 1`] = `
<DocumentFragment>
  <div
    aria-label="1"
    class="wrapper isCurrent isSelected isHighlighted"
    role="button"
    tabindex="0"
  >
    1
  </div>
</DocumentFragment>
`;

exports[`Year display props it renders 2`] = `
<DocumentFragment>
  <div
    aria-label="1"
    class="wrapper isCurrent isSelected isHighlighted"
    role="button"
    tabindex="0"
  >
    1
  </div>
</DocumentFragment>
`;

exports[`Year display props it renders 3`] = `
<DocumentFragment>
  <div
    aria-label="1"
    class="wrapper"
    role="button"
    tabindex="0"
  >
    1
  </div>
</DocumentFragment>
`;
